File: hearts-scala-mike/src/main/scala/funar/hearts/Table.scala
package funar.hearts

import scala.annotation.tailrec

object Table {
  // should only call this with 4 cards
  // non-total function
  def whoTakesTrick(trick: Trick): Player = {
    assert(!trick.isEmpty)
    @tailrec
    def loop(player: Player, card: Card, trick: Trick): Player =
      trick match {
        case Nil => player
        case ((player1, card1)::rest) =>
          card.beats(card1) match {
            case None => loop(player, card, rest)
            case Some(false) => loop(player, card, rest)
            case Some(true) => loop(player1, card1, rest)
          }
      }
    val (player0, card0) :: rest1 = trick.reverse
    loop(player0, card0, rest1)
  }

  def legalCard(card: Card, hand: Hand, trick: Trick): Boolean =
    hand.contains(card) &&
    (Trick.isEmpty(trick) ||
     {
       val firstCard = Trick.leadingCard(trick)
       val firstSuit = firstCard.suit
       (card.suit == firstSuit || hand.forall(_.suit != firstSuit))
     })

  def cardScore(card: Card): Integer =
    card match {
      case Card(Suit.Spades, Rank.Queen) => 13
      case Card(Suit.Hearts, _) => 1
      case _ => 0
    }

  // cards taken from trick
  type Pile = Set[Card]

  type PlayerPiles = Map[Player, Pile]

  // projection / query model
  case class TableState(players: List[Player],  // player whose turn it is is at the front
                        hands: PlayerHands,
                        piles: PlayerPiles,
                        trick: Trick)

  def emptyTableState(players: List[Player]): TableState =
    TableState(players = players,
               hands = Map.from(players.map((_, Hand.empty))),
               piles = Map.from(players.map((_, Set.empty))),
               trick = Trick.empty)

  def gameAtBeginning(tableState: TableState): Boolean =
    Trick.isEmpty(tableState.trick) &&
      tableState.piles.values.forall(_.isEmpty)

  def playerAfter(tableState: TableState, player: Player): Player =
    rotate(rotateTo(player, tableState.players)).head

  def rotate[A](list: List[A]): List[A] = {
    assert(!list.isEmpty)
    list.tail ++ List(list.head)
  }

  @tailrec
  def rotateTo[A](x: A, list: List[A]): List[A] = {
    assert(!list.isEmpty)
    if (list.head == x)
      list
    else
      rotateTo(x, rotate(list))
  }

  def currentPlayer(tableState: TableState): Player =
    tableState.players.head

  def playValid(tableState: TableState, player: Player, card: Card): Boolean = {
    val hand = tableState.hands(player)
    System.out.println("playValid " + tableState + " " + player + " " + hand)
    val trick = tableState.trick
    legalCard(card, hand, trick) &&
     (if (gameAtBeginning(tableState))
        card == Card(Suit.Clubs, Rank.Two)
      else
        currentPlayer(tableState) == player)
  }

  // turn: 4 cards on the trick
  def turnOver(tableState: TableState): Boolean =
    tableState.players.length == tableState.trick.length


  def pileScore(pile: Pile): Integer =
    pile.map(cardScore).foldLeft(0)(_ + _)

  def gameOver(tableState: TableState): Option[Player] =
    if (tableState.hands.values.forall(_.isEmpty)) {
      val playerScores = tableState.piles.map { case (player, pile) => (player, pileScore(pile)) }
      Some(playerScores.minBy(_._2)._1)
    } else 
      None

  def takeCard(hands: PlayerHands, player: Player, card: Card): PlayerHands =
    hands.updatedWith(player) { o => o.map(_ - card) }

  def addToPile(playerPiles: PlayerPiles, player: Player, cards: Seq[Card]): PlayerPiles = {
    val playerPile = playerPiles.getOrElse(player, Set.empty)
    playerPiles + (player -> playerPile.union(Set.from(cards)))
  }

  def tableProcessEvent(event: GameEvent, tableState: TableState): TableState =
    event match {
      case GameEvent.HandDealt(player, hand) => 
        tableState.copy(hands = tableState.hands + (player -> hand),
                        trick = Trick.empty)
      case GameEvent.PlayerTurnChanged(player) =>
        tableState.copy(players = rotateTo(player, tableState.players))
      case GameEvent.LegalCardPlayed(player, card) =>
        tableState.copy(hands = takeCard(tableState.hands, player, card),
                        trick = Trick.add(tableState.trick, player, card))
      case GameEvent.IllegalCardPlayed(player, card) => tableState
      case GameEvent.TrickTaken(player, trick) =>
        tableState.copy(piles = addToPile(tableState.piles, player, Trick.cards(trick)),
                        trick = Trick.empty)
      case GameEvent.GameEnded(player) => tableState
    }


  def tableProcessCommand(command: GameCommand, tableState: TableState): Seq[GameEvent] =
    command match {
      case GameCommand.DealHands(hands) =>
        hands.map(GameEvent.HandDealt.tupled).toSeq
      case GameCommand.PlayCard(player, card) =>
        if (playValid(tableState, player, card)) {
          val event1 = GameEvent.LegalCardPlayed(player, card)
          val tableState1 = tableProcessEvent(event1, tableState)
          if (turnOver(tableState1)) {
            val trickTaker = whoTakesTrick(tableState1.trick)
            val event2 = GameEvent.TrickTaken(trickTaker, tableState1.trick)
            val tableState2 = tableProcessEvent(event2, tableState1)
            gameOver(tableState2) match {
              case None => Seq(event1, event2, GameEvent.PlayerTurnChanged(trickTaker))
              case Some(winner) =>
                Seq(event1, event2, GameEvent.GameEnded(winner))
            }
          } else
            Seq(event1, GameEvent.PlayerTurnChanged(playerAfter(tableState1, player)))
        } else
          Seq(GameEvent.IllegalCardPlayed(player, card))
    }
}

File: johannes/hearts/src/Table.hs
module Table(TableState, emptyTableState,
             runTable)
where

import qualified Data.Foldable as Foldable

import qualified Data.Set as Set
import Data.Set (Set)

import qualified Data.Map.Strict as Map
import Data.Map.Strict (Map, (!))

import Cards
import GameEvent

import Debug.Trace (trace)

-- ist die Karte aus der Hand für den Stich zulässig ?
legalCard :: Card -> Hand -> Trick -> Bool
legalCard card hand trick =
  containsCard card hand &&
  (trickEmpty trick ||
   (let firstCard = leadingCardOfTrick trick
        firstSuit = suit firstCard
    in  suit card == firstSuit -- ok if suit is followed
        || all ((/= firstSuit) . suit) (handCards hand))) -- ok if no such suit in hand

-- Wert einer Karte
cardScore :: Card -> Integer
cardScore (Card Spades Queen) = 13
cardScore (Card Hearts _) = 1
cardScore _ = 0

-- * Spiellogik

-- Hände der Spieler
type PlayerHands = Map Player Hand

-- eingezogene Karten, pro Spieler
type PlayerPiles = Map Player Pile

data TableState =
  TableState
  { tableStatePlayers :: [Player], -- wer dran ist, steht vorn
    tableStateHands   :: PlayerHands,
    tableStatePiles  :: PlayerPiles,
    tableStateTrick   :: Trick
  }
  deriving Show

-- Anfangszustand herstellen
emptyTableState :: [Player] -> TableState
emptyTableState players =
  TableState {
    tableStatePlayers = players,
    tableStateHands = Map.fromList (map (\ player -> (player, emptyHand)) players),
    tableStatePiles = Map.fromList (map (\ player -> (player, emptyPile)) players),
    tableStateTrick = emptyTrick
  }

-- ist das Spiel noch am Anfang?
gameAtBeginning :: TableState -> Bool
gameAtBeginning tableState =
  (trickEmpty (tableStateTrick tableState)) && 
    (all pileEmpty (Map.elems (tableStatePiles tableState)))

-- wer ist als nächstes dran?
playerAfter :: TableState -> Player -> Player
playerAfter state player =
   head (rotate (rotateTo player (tableStatePlayers state)))

-- Liste rotieren
rotate :: [a] -> [a]
rotate (x : xs) = xs ++ [x]
rotate [] = undefined

-- Liste zu einem bestimmten Element rotieren
rotateTo :: Eq a => a -> [a] -> [a]
rotateTo y xs@(x : xs') | x == y = xs
                        | otherwise = rotateTo y (xs' ++ [x])
rotateTo y [] = undefined


-- wer ist gerade dran?
currentPlayer :: TableState -> Player
currentPlayer state =
  head (tableStatePlayers state)


-- ist es zulässig, diese Karte auszuspielen?
playValid :: TableState -> Player -> Card -> Bool
playValid tableState player card =
  let hand = tableStateHands tableState ! player
      trick = tableStateTrick tableState
  in
  legalCard card hand trick &&
  if gameAtBeginning tableState
  then card ==  Card Clubs Two
  else currentPlayer tableState == player

-- ist diese Runde vorbei?
turnOver :: TableState -> Bool
turnOver state =
  length (tableStatePlayers state) == length (trickToList (tableStateTrick state))

-- wer muß den Stich einziehen?
whoTakesTrick :: Trick -> Maybe Player
whoTakesTrick (Trick []) = Nothing
whoTakesTrick (Trick list) =
  case reverse list of 
    [] -> Nothing
    ((player0, card0) : rest') ->
      let loop player _ [] = Just player
          loop player card ((player', card') : rest) =
            case cardBeats card' card of
              Nothing -> loop player card rest
              Just False -> loop player card rest
              Just True -> loop player' card' rest
      in loop player0 card0 rest'

turnOverTrick :: TableState -> Maybe (Trick, Player)
turnOverTrick state =
  if turnOver state
  then
    let trick = tableStateTrick state
    in fmap (\player -> (trick, player)) (whoTakesTrick trick)
  else Nothing

-- Wert eines Stapels
pileScore :: Pile -> Integer
pileScore pile = sum (map cardScore (pileCards pile))

-- Ist das Spiel vorbei und wenn ja wer hat gewonnen?
gameOver :: TableState -> Maybe Player
gameOver state =
  if all isHandEmpty (Map.elems (tableStateHands state))
  then 
    let playerScores = Map.toList (fmap pileScore (tableStatePiles state))
        cmp (_, score1) (_, score2) = compare score1 score2
    in Just (fst (Foldable.minimumBy cmp playerScores))
  else 
    Nothing

-- Karte ausspielen
playCard :: PlayerHands -> Player -> Card -> PlayerHands
playCard playerHands player card =
  Map.alter (fmap (removeCard card)) player playerHands

-- Karten zum Stapel hinzufügen
addTrickToPile :: PlayerPiles -> Player -> Trick -> PlayerPiles
addTrickToPile playerPiles player trick =
  let playerPile = Map.findWithDefault emptyPile player playerPiles
  in Map.insert player (pileAddTrick playerPile trick) playerPiles

dealHand :: Player -> Hand -> PlayerHands -> PlayerHands
dealHand player hand hands = Map.insert player hand hands

-- Ereignis in den Zustand einarbeiten
tableProcessEvent :: GameEvent -> TableState -> TableState
-- processGameEvent event state | trace ("processGameEvent " ++ show state ++ " " ++ show event) False = undefined
tableProcessEvent (HandDealt player hand) state =
  state {
    tableStateHands = dealHand player hand (tableStateHands state)
  }
tableProcessEvent (PlayerTurnChanged player) state =
  state {
    tableStatePlayers  = rotateTo player (tableStatePlayers state)
  }
tableProcessEvent (LegalCardPlayed player card) state =
  state {
    tableStateHands = playCard (tableStateHands state) player card,
    tableStateTrick = addToTrick player card (tableStateTrick state)
  }
tableProcessEvent (TrickTaken player trick) state =
  state {
    tableStatePiles =
      addTrickToPile (tableStatePiles state) player trick,
    tableStateTrick = emptyTrick
  }
tableProcessEvent (IllegalCardAttempted player card) state = state
tableProcessEvent (GameEnded player) state = state

-- Spielablauf interpretieren/ausführen
runGame :: Game a -> TableState -> (TableState, a)
runGame (IsCardValid player card callback) state =
  runGame (callback (playValid state player card)) state
runGame (TurnOverTrick callback) state =
  runGame (callback (turnOverTrick state)) state
runGame (PlayerAfter player callback) state =
  runGame (callback (Table.playerAfter state player)) state
runGame (IsGameOver callback) state =
  runGame (callback (gameOver state)) state
runGame (Return result) state =
  (state, result)
runGame (RecordEvent event callback) state =
  runGame (callback ()) (tableProcessEvent event state)
runGame _ _ = undefined

runTable = runGame